### ISSUE TITLE
Did you have the performance summary of the RL-based quantization performance on MobileNet?

### ISSUE THREAD

**Issue by brisker** [Tencent/PocketFlow#115]
Did you have the performance summary of the RL based quantization performance on MobileNet? Given that you have described about this here :  https://pocketflow.github.io/reinforcement_learning/

**Comment by haolibai** [comment]
Yes. We have some preliminary results on MobileNet, with `uql_evquivalent_bits=4` or `nuql_evquivalent_bits=4`. ( 8 bit quantization does not hurt the performance, while 2 bit degrade the performance significantly).

For `uql_evquivalent_bits=4`, the top-1 acc is 47.1%~53.0%.
For `nuql_evquivalent_bits=4`, the top-1 acc is 64.5%~65.9%.

The accuracy on validation set is measured at a fixed interval. Note that the variance of acc could be large sometimes. The above results may be improved if hyper-parameters of the RL agent are more carefully designed.

**Comment by brisker** [comment]
@haolibai 
Why is the non-uniform quantization performance much better than uniform quantization? In standard quantization methods, it seems to be not like this.

**Comment by haolibai** [comment]
Generally non-uniform quantization could have larger model capacity and thereon can outperform uniform quantization, especially in low-bit settings.

**Comment by brisker** [comment]
How to reproduce the RL-based quantization performance?
which scripts?
@haolibai

**Comment by haolibai** [comment]
The model is evaluated every 10000 steps, so there could be some variance on the accuracy. We found that for mobilenet, the variance is a little bit large (e.g., 47.1%~53.0%).
Remain the default parameters for the RL agent, i.e., run
./scripts/run_seven.sh nets/mobilenet_at_ilsvrc12_run.py \
    --learner uniform \
    --uql_enbl_rl_agent

**Comment by brisker** [comment]
@haolibai 
1. the performance of MobileNet, is it Mobilenet v1 or v2?
2. What is the 32-bit precision performance of the v1 and v2 mobilenet model, downloaded from https://api.ai.tencent.com/pocketflow  ??

**Comment by haolibai** [comment]
1. Mobilenet v1
2. The full precision results are in https://pocketflow.github.io/pre_trained_models/
Downloaded from: https://api.ai.tencent.com/pocketflow/models_mobilenet_v1_at_ilsvrc_12.tar.gz

**Comment by brisker** [comment]
@haolibai 
I have run this command:(with default settings)
`bash ./scripts/run_local.sh nets/mobilenet_at_ilsvrc12_run.py -n=4 --learner=uniform --uql_enbl_rl_agent=True --uql_equivalent_bits=4`, and in the first several iterations, the acc is so low(see the pic below). Is this normal?
![snipaste_2018-12-03_19-52-50](https://user-images.githubusercontent.com/13804492/49373190-1063b800-f738-11e8-90eb-f7616abd6d8c.png)

**Comment by brisker** [comment]
@jiaxiang-wu

**Comment by brisker** [comment]
@haolibai 
But I just follow the default settings. ...
Did you mean the default settings have bugs?....

**Comment by haolibai** [comment]
The first 25% epochs are just for collecting training data for the buffer of DDPG, while the RL agent has not started training. The searching space is a little bit large, and MobileNet-v1 is indeed sensitive to low-bit quantization, so in most cases if the strategy is not good enough, the result will be pretty poor.

If I remember correctly, I generate the above results using the default settings. Nevertheless, the default setting may be not the best setting for the training.

**Comment by brisker** [comment]
@haolibai 
Besides, I have used 4 gpus, but when running `nvidia-smi`, the actual gpu-util of the last 3 gpus is actually 0%. So does the RL-based quantization support multi-gpu training?

**Comment by haolibai** [comment]
The training of RL agent will use only one GPU, while for training the network parameters will use 4 GPUs.
Please check if you successfully use 4 GPUs when using other learners?

**Comment by brisker** [comment]
@haolibai 
Currently I am only using the RL-based UniformQuantization learner. The mobilenet-v1 RL-based model has run over 120 roll-out, and the acc is still nearly 0. I just follow the default settings with the following command:
`bash ./scripts/run_local.sh nets/mobilenet_at_ilsvrc12_run.py -n=4 --learner=uniform --uql_enbl_rl_agent=True --uql_equivalent_bits=4`
I am wondering if the version of the code has some changes, given that you mentioned you have got a 47.1%~53.0% performance with mobilenet-v1.

**Comment by haolibai** [comment]
Did you set `enbl_warm_start=True`? Since the quantized network is based on a pretrained model.
Sorry this should be added in the tutorial.

**Comment by brisker** [comment]
previously no. But just a moment ago, as you have mentioned, I added this flag into the command, and run it, but still, in the first several iterations, the acc is still nearly 0. It seems to be no difference. So is there any other possible mistakes?
@haolibai

**Comment by haolibai** [comment]
I tried a moment ago, using the following configurations, and there are effective accs.

./scripts/run_seven.sh nets/resnet_at_ilsvrc12_run.py -n=8 \
--learner uniform \ 
--enbl_warm_start \
--uql_enbl_rl_agent \
--uql_equivalent_bits 4 \ 
--uql_w_bit_min 2 \
--uql_w_bit_max 8

![image](https://user-images.githubusercontent.com/24502378/49429610-dd76fe00-f7e3-11e8-90c2-fa68e3bcf195.png)
(seven is just a distributed system, you can replace it with your own GPUs)
Please make sure that you have successfully restored a pretrained MobileNet-v1, and correct hyper-parameters.

**Comment by brisker** [comment]
@haolibai 
but what I refer to is mobilenet_v1, and the command I run is 
`bash ./scripts/run_local.sh nets/mobilenet_at_ilsvrc12_run.py -n=4 --learner=uniform --uql_enbl_rl_agent=True --uql_equivalent_bits=4`

**Comment by jiaxiang-wu** [comment]
@brisker The key is that you need to enable warm-start by specifying `--enbl_warm_start`.

**Comment by haolibai** [comment]
The following are just the default settings, you can ignore them:
--uql_w_bit_min 2
--uql_w_bit_max 8

**Comment by brisker** [comment]
@haolibai 
@jiaxiang-wu 
already changing the command to  
`bash ./scripts/run_local.sh nets/mobilenet_at_ilsvrc12_run.py -n=4 --learner=uniform --uql_enbl_rl_agent=True --uql_equivalent_bits=4 --enbl_warm_start=True`
still nearly 0 acc.

**Comment by brisker** [comment]
@haolibai 
changing the model to resnet18 , still nearly 0 acc.

**Comment by brisker** [comment]
@haolibai 
now it is normal. thanks. Sorry for bothering.

**Comment by brisker** [comment]
@haolibai 
The problem seems not to be totally solved. I noticed that in the training process, the layer_bits are often allocated a lot of 2-bits , even I have assigned uql_equivalent_bits=4. (But in your log I can see many bits above 2) And I assum this is the reason why the acc is always 0. Here is my log. Could you please check anything wrong about the layer_bits allocation during training?


[resnet18_imagenet_4bit_log.log](https://github.com/Tencent/PocketFlow/files/2643838/resnet18_imagenet_4bit_log.log)

**Comment by jiaxiang-wu** [comment]
@brisker Potential bug. We are looking at this issue.

**Comment by brisker** [comment]
besides, If I run the resnet_cifar experiments, the acc_top1 is normal, but the acc_top5 is all 0. 
I assume this is a minor bug.
@haolibai 
@jiaxiang-wu

**Comment by jiaxiang-wu** [comment]
@brisker We noticed that the accuracy recovers quickly after the RL-based hyper-parameter optimization stage finishes and the final re-training process starts. Maybe something went wrong in the fast fine-tuning code for RL-based hyper-parameter optimization. Anyway, thanks for your feedback.

**Comment by brisker** [comment]
@jiaxiang-wu 
but if I run the resnet_cifar experiment, although the printed acc_top5 is all 0, but the acc_top1 is very high, around 85%~100%.

**Comment by jiaxiang-wu** [comment]
@brisker 
I think it should the latter, that CIFAR-10 is too easy so that some bug does not lead to dramatically reduction in the top-1 accuracy.

**Comment by brisker** [comment]
@jiaxiang-wu 
I have not carefully read the code. But if there exists too many 0-acc in the RL process, the RL seems to lose its meaning. Previously @haolibai  told me that he can get about 50% acc for mobilenet-imagenet, I am curious about that, if there still exists many 0-acc in his log.

**Comment by jiaxiang-wu** [comment]
@brisker 
Maybe some recent code modification introduces bugs. If there are too many 0-acc in the RL process, the RL agent is likely to fail in finding the optimal hyper-parameter combination.

**Comment by brisker** [comment]
@jiaxiang-wu  So can you tell me which version of Pocketflow can you reproduce the RL-based quantization performance, as @haolibai  claimed?

**Comment by haolibai** [comment]
It is an earlier version before the release. Let us reproduce the result and inform you later.

**Comment by brisker** [comment]
@haolibai 
what is the difference between `bucket_type==split and bucket_type==channel`? 
what is the type in the experiments you have reproduced(the mobilenet 47.1%~53.0%. performance )?

**Comment by haolibai** [comment]
Without bucketing.

**Comment by brisker** [comment]
given a conv layer with `[k,k,cin,cout] `weights , and `N=k*k*cin*cout` represents float values, 
if without bucketing, all the `N` values are quantized using the same alpha and beta.
if with bucketing with `bucket_type==channel`, all the `N `values are quantized using `cout` alphas and `cout` betas?
Have I understood this right?
@haolibai

**Comment by haolibai** [comment]
Yes correct.

**Comment by brisker** [comment]
@haolibai 
If `--uql_enbl_rl_agent ==True,`  the weight quantization bits is learned, what about the activation bits?
`32`? or `equivalent_bits`?

**Comment by haolibai** [comment]
`activation_bits` remains 32 bits, since quantization with different bits across layers can not be accelerated, which poses no necessity for activation quantization.

**Comment by brisker** [comment]
@haolibai 
I do not quite understand what you mean.
Why the different activation bit across layers can not be  accelerated?
here https://github.com/haolibai/PocketFlow/blob/master/learners/uniform_quantization/bit_optimizer.py#L52
it says `Currently only weight bits are inferred via RL. Activations later.` 
So what is your plan in the future, regarding the activation-RL?
@haolibai 
@jiaxiang-wu

**Comment by brisker** [comment]
I do not quite understand what you mean.
Why the different activation bit across layers can not be accelerated?
here https://github.com/haolibai/PocketFlow/blob/master/learners/uniform_quantization/bit_optimizer.py#L52
it says Currently only weight bits are inferred via RL. Activations later.
So what is your plan in the future, regarding the activation-RL?
@haolibai
@jiaxiang-wu

**Comment by haolibai** [comment]
A basic requirement for low-bit inference on most current deep learning framework is that you need to maintain the same precision of bits for all the data. If weights are quantized with different bits, the requirement cannot be satisfied, and thereon cannot be accelerated. Nevertheless, weight quantization still can reduce the model size. However, for activation quantization, it can neither reduce the model size nor be accelerated, which brings no gain. 
Of course we can do RL based quantization on activations theoretically, but only when it can bring some practical advantages will we implement it.

**Comment by brisker** [comment]
@haolibai 
@jiaxiang-wu 
Will  Pocketflow support RL-agent reward signal directly from hardware?

**Comment by haolibai** [comment]
I am not sure.

**Comment by brisker** [comment]
@haolibai 
so until now, is there bugs found in the quantization code, regarding the RL-based method?

**Comment by brisker** [comment]
@haolibai 
It seems that there is no corresponding paper about the RL-based quantization part. Could you please provide any documention about the RL-based quantization algorithm? Or any related paper?

**Comment by haolibai** [comment]
We refer to http://openaccess.thecvf.com/content_ECCV_2018/papers/Yihui_He_AMC_Automated_Model_ECCV_2018_paper.pdf to realize the RL based quantization. Please check.

**Comment by brisker** [comment]
@haolibai 
@jiaxiang-wu 
1. Is this pic corresponding to the RL-based quantization performance for imagenet-resnet18?
2. Did this pic mean that w4a32 quantization model with no RL only has about 63% top1-accuracy on imagenet, and with RL the acc can rise up to 68?
3.  Are these performance with channel bucket?

**Comment by haolibai** [comment]
1. Yes.
2. In later experiment we found that w4a32 without RL can also achieve acc up to 66~67 by decreasing the learning rate properly. The RL based results have some variance, but can generally achieve comparable or better results.
3. Without bucketing.

**Comment by brisker** [comment]
@haolibai 
why is the lr always very very small(usually) , no matter it is RL-based method or not.  In the log you sent me [here](https://github.com/Tencent/PocketFlow/issues/126#issuecomment-444805789), the learning rate start from 2e-6. Is it necessary to set it as small as 2e-6 ? Or the printed `2e-6` is not the "real " learning rate?

**Comment by brisker** [comment]
@haolibai 

 In resnet-imagenet RL experiment, the acc may be normal. But when it comes to mobilenet, the reward got too many 0 in the roll-out process. I think it is due to the uniform quantization algorithm itself ( it is not a state-of-art model quantization paper. Is it a paper from Google?). So here I am curious about the performance you mentioned [here](https://github.com/Tencent/PocketFlow/issues/115#issuecomment-443198659). Is there also many 0 acc during the roll-out process  in the experiment there?

**Comment by brisker** [comment]
@haolibai 
Now I can also reproduce the resnet-18-4bit rl-based experiment performance, around 67% top-acc, but the mobilenet-rl-based experiment just still got too many 0 accs in the roll-out training process. Have you reproduce this , except the resnet experiment [here](https://github.com/Tencent/PocketFlow/issues/115#issuecomment-443198659)  ?

**Comment by haolibai** [comment]
Hi. Not yet. We are recently short of GPUs. We will inform you immediately once we get the result. You can also try to adjust some parameters, such as the learning rate for RL global training, and narrow down the gap between w_bit_min and w_bit_max so as to avoid too many 2 bit quantizations.

**Comment by brisker** [comment]
@haolibai 
in the log you provided [here](https://github.com/Tencent/PocketFlow/issues/126#issuecomment-444805789) , in the last line : 
```
INFO:tensorflow:accuracy: 0.6690799593925476
```
does these two lines mean that the **top-1 acc** is 0.669? or **top-5 acc**?

**Comment by brisker** [comment]
@haolibai 
@jiaxiang-wu 
Is the potential bug of RL-based quantization code solved?
(I run the mobilenet-8-8 uniform quantization with no rl and no distillation, and the performance is about 69.5, which is 1.5 point lower than what you claimed in the readme file)

**Comment by brisker** [comment]
@haolibai
@jiaxiang-wu
Is some bug found now, regarding to the RL-quantization code?
BTW, 
1. I found a bug in the code here: https://github.com/Tencent/PocketFlow/blob/master/learners/uniform_quantization/learner.py#L279
I think it should be 
`images.set_shape((FLAGS.batch_size_eval, images.shape[1], images.shape[2],images.shape[3]))`
not
`images.set_shape((FLAGS.batch_size, images.shape[1], images.shape[2],images.shape[3]))`
2. Currently I found the resnet-imagenet reward is normal, but the mobilenetv1-imagenet reward is all the way very low (around below 0.1)
during all the roll-out process, but the during roll-outs the training acc is relatively high(around 0.5), I do not know if this is a bug too.

**Comment by brisker** [comment]
@haolibai 
@jiaxiang-wu 
It has been more than 2 weeks.....

**Comment by jiaxiang-wu** [comment]
Yes, it should be `FLAGS.batch_size_eval` instead of `FLAGS.batch_size`. Thanks for pointing it out.
For the potential bug in the RL-quantizaiton code, our team does not have enough people to cover that, at least for the moment. If it is urgent, you can try to debug it by yourself.
@brisker

**Comment by brisker** [comment]
@jiaxiang-wu 
when will the 2.0 version be released? what will be new features?

**Comment by jiaxiang-wu** [comment]
@brisker 
Possibly at the end of 2019Q1. We are now adding support for object detection models (*e.g.* SSD) and RNN/LSTM models.

**Comment by xiaomr** [comment]
@brisker @haolibai  Have you reproduce the mobilenet-imagenet-rl-experiment , as reported [here](https://github.com/Tencent/PocketFlow/issues/115#issuecomment-443198659),  I folowed the default scripts and encountered the same acc 0 issue as @brisker
